.github/workflows/pypi-release.yml

```diff
@@ -2,157 +2,211 @@
-  
-  #   environment: PypiTest
-  #   steps:
-  #     - uses: actions/download-artifact@v3
-  #       with:
-  #         name: dist
-  #         path: 'dist'
-
-  #     - name: Publish distribution to test PyPI
-  #       uses: pypa/gh-action-pypi-publish@v1.5
-  #       with:
-  #         password: ${{ secrets.TEST_PYPI_API_TOKEN }}
-  #         repository_url: https://test.pypi.org/legacy/
-
-  # prod-pypi-release:
-  #   runs-on: ubuntu-latest
-  #   needs: [check-pypi-exists]
-  #   if: needs.check-pypi-exists.outputs.exists == 'false'  && ${{ inputs.testing }} == 'true'
-  
-  #   environment: PypiProd
-  #   steps:
-  #     - uses: actions/download-artifact@v3
-  #       with:
-  #         name: dist
-  #         path: 'dist'
-
-  #     - name: Publish distribution to PyPI
-  #       if: ${{ inputs.testing }} == 'true'
-  #       uses: pypa/gh-action-pypi-publish@v1.5
-  #       with:
-  #         password: ${{ secrets.PYPI_API_TOKEN }}
-
-  # validate-pypi-package:
-  #   runs-on: ubuntu-latest
-  #   needs: [test-pypi-release, prod-pypi-release]
-  #   # always run this step because one of the needs are always skipped.
-  #   if: always()
-
-  #   steps:
-  #     - name: Check PyPI
-  #       id: pypi_info
-  #       uses: dbt-labs/actions/py-package-info@v1
-  #       with:
-  #         package: ${{ github.event.repository.name }}
-  #         version: ${{ inputs.tag }}
-  #         # test: ${{ inputs.testing }}  TODO: implement in action
-      
-  #     - name: Set output
-  #     # The above step will just use the latest version if the input version
-  #     # is not found.  So to validate the version we want to release exists
-  #     # we need to compare teh output version.
-  #       id: set_existence
-  #       run: |
-  #         if [[ ${{ steps.pypi_info.version }} != ${{ inputs.tag }} ]]
-  #         then
-  #           echo ${{ inputs.tag }} released to Pypi
-  #         else
-  #           echo ${{ inputs.tag }} FAILED to released to Pypi.  Manual intervention required.
-  #           exit 1
-  #         fi
+      - name: "[Notification] Package Version Already Live In Package Index. Skip Upload."
+        run: |
+          title="Package version already live in package index"
+          version=${{ inputs.version_number }}
+          package_name=${{ needs.sanitize-package-name.outputs.name }}
+          package_index="PyPI"
+          if [[ ${{ inputs.test_run }} == true ]]
+          then
+            package_index="Test PyPi"
+          fi
+          message="The version $version of package $package_name already lives in the $package_index. The upload to the package index will be skipped."
+          echo "::notice title=${{ env.NOTIFICATION_PREFIX }}: $title::$message"
+
+  test-pypi-release:
+    runs-on: ubuntu-latest
+    needs: [sanitize-package-name, check-package-exists-pypi]
+    if: >-
+      needs.check-package-exists-pypi.outputs.exists == 'false' &&
+      inputs.test_run == 'true'
```

Comment:
Is it accurate that booleans are always treated as strings for all input contexts with a trigger of `workflow_call`.

But boolean inputs are treated as true booleans (not strings) for the `inputs` context of a `workflow_dispatch` trigger only?  And still treated as strings for the `github.events.inputs` context?

Reply:
Inputs contexts for `workflow_dispatch` and `workflow_call` were unified ([_Related announcement_](https://github.blog/changelog/2022-06-10-github-actions-inputs-unified-across-manual-and-reusable-workflows/)). According to [documentation](https://docs.github.com/en/actions/learn-github-actions/contexts#inputs-context) the input context provides typed data. This means we must respect declared types if we are trying to access the workflow input data via the input context. In case when we refer to input through GitHub context (`github.events.inputs`), all data is cast to string type.

Regarding the outputs, the following contexts store and provide data as a string:
- `jobs.<job_id>.outputs.<output_name>` ([`jobs` context](https://docs.github.com/en/actions/learn-github-actions/contexts#jobs-context));
- `steps.<step_id>.outputs.<output_name>` ([`steps` context](https://docs.github.com/en/actions/learn-github-actions/contexts#steps-context));
- `needs.<job_id>.outputs.<output name>` ([`needs` context](https://docs.github.com/en/actions/learn-github-actions/contexts#needs-context));